### PR TITLE
Optimize with IN Clause for UPDATE/DELETE Statements on Vindexes

### PR DESCRIPTION
**Pull request by wangweicugw**
## Description

When I run a SELECT statement using an IN clause with the Vindexes, the query is correctly split according to the corresponding values in the IN list. This ensures that each shard is queried separately, which is the expected behavior for optimized shard querying.

However, when I perform an UPDATE/DELETE statement with an IN clause on the Vindexes, the SQL is not rewritten in the same manner. Instead, the query is sent directly to the shard without being split based on the IN clause values.

I hope to have consistent behavior with SELECT when processing UPDATE/DELETE statements with an IN clause on the Vindexes.

I would like the following UPDATE statement:
```sql
UPDATE test_vindex_signed_int SET id = id + 100 WHERE v_key IN (1, 100)
```
to be rewritten into separate statements, one for each value in the IN clause, like so:
```sql
UPDATE test_vindex_signed_int SET id = id + 100 WHERE v_key IN (1)
UPDATE test_vindex_signed_int SET id = id + 100 WHERE v_key IN (100)
```
This would ensure that the UPDATE/DELETE operation is as efficient as possible, especially for large batch updates, by targeting only the relevant shards.

## Related Issue(s)
- Closes https://github.com/vitessio/vitess/issues/15453
- Closes https://github.com/vitessio/vitess/issues/6343

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 1]
```diff
@@ -771,7 +771,9 @@ func updateSelectedVindexPredicate(op *operators.Route) sqlparser.Expr {
 		if !ok {
 			continue
 		}
-
+		if sqlparser.Equals.Expr(cmp.Right, sqlparser.ListArg(engine.DmlVals)) {
+			continue
+		}
```
Comment (wangweicugw) [inline go/vt/vtgate/planbuilder/operator_transformers.go:776]: I think there can be a better solution for this.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:776]: This is a very fragile hack.

**Comment by systay** [comment]
@wangweicugw Nice stuff!

Did you check that we have relevant end-to-end tests for this?

File: go/vt/vtgate/engine/dml.go [hunk 2]
```diff
@@ -78,7 +78,7 @@ func (dml *DML) execUnsharded(ctx context.Context, primitive Primitive, vcursor
 	return execShard(ctx, primitive, vcursor, dml.Query, bindVars, rss[0], true /* rollbackOnError */, !dml.PreventAutoCommit /* canAutocommit */)
 }
 
-func (dml *DML) execMultiDestination(ctx context.Context, primitive Primitive, vcursor VCursor, bindVars map[string]*querypb.BindVariable, rss []*srvtopo.ResolvedShard, dmlSpecialFunc func(context.Context, VCursor, map[string]*querypb.BindVariable, []*srvtopo.ResolvedShard) error) (*sqltypes.Result, error) {
+func (dml *DML) execMultiDestination(ctx context.Context, primitive Primitive, vcursor VCursor, bindVars map[string]*querypb.BindVariable, rss []*srvtopo.ResolvedShard, dmlSpecialFunc func(context.Context, VCursor, map[string]*querypb.BindVariable, []*srvtopo.ResolvedShard) error, bvs []map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
```
Comment (systay) [inline go/vt/vtgate/engine/dml.go:81]: nitpick: this line is so long it should probably be split up

File: go/vt/vtgate/engine/update_test.go [hunk 3]
```diff
@@ -949,6 +949,33 @@ func TestUpdateMultiEqual(t *testing.T) {
 	})
 }
 
+func TestUpdateInUnique(t *testing.T) {
+	ks := buildTestVSchema().Keyspaces["sharded"]
```
Comment (systay) [inline go/vt/vtgate/engine/update_test.go:953]: a short description of what you are trying to test would be helpful here

**Comment by wangweicugw** [comment]
> @wangweicugw Nice stuff!
> 
> Did you check that we have relevant end-to-end tests for this?

I haven't found relevant end-to-end tests for this; I will supplement them in go/test/endtoend/vtgate/queries/dml/dml_test.go.

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:776]: We can ignore if right is an argument type

**Review by harshit-gangal — COMMENTED** [review]
I could not find any test that shows the bind variables are split across different shards for `__vals`

**Comment by harshit-gangal** [comment]
> > @wangweicugw Nice stuff!
> > Did you check that we have relevant end-to-end tests for this?
> 
> I haven't found relevant end-to-end tests for this; I will supplement them in go/test/endtoend/vtgate/queries/dml/dml_test.go.

you should also validate this in a unit test that shows what bind variables were sent down.

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 1 above)
Comment (wangweicugw) [reply go/vt/vtgate/planbuilder/operator_transformers.go:776]: > We can ignore if right is an argument type

I tried making this modification, 
```
func updateSelectedVindexPredicate(routing operators.Routing) sqlparser.Expr {
	tr, ok := routing.(*operators.ShardedRouting)
	if !ok || tr.Selected == nil {
		return nil
	}

	_, isMultiColumn := tr.Selected.FoundVindex.(vindexes.MultiColumn)
	for idx, expr := range tr.Selected.Predicates {
		cmp, ok := expr.(*sqlparser.ComparisonExpr)
		if !ok || cmp.Operator != sqlparser.InOp {
			continue
		}
		_, ok = cmp.Left.(*sqlparser.ColName)
		if !ok {
			continue
		}
		_, ok = cmp.Right.(sqlparser.ListArg)
		if ok {
			continue
		}
		var argName string
		if isMultiColumn {
			argName = engine.ListVarName + strconv.Itoa(idx)
		} else {
			argName = engine.ListVarName
		}

		cmp.Right = sqlparser.ListArg(argName)
	}
	return nil
}
```

but it caused some other cases in plan_test to fail, just like this one.
```
        Diff:
        {
            "Instructions": {
                "Inputs": [
                    {
                        "Aggregates": "any_value(0) AS 1, sum_count(1) AS count(id)",
                        "Inputs": [
                            {
                                "Inputs": [
                                    {
                                        "FieldQuery": "select `name`, keyspace_id from name_user_vdx where 1 != 1",
                                        "Keyspace": {
                                            "Name": "user",
                                            "Sharded": true
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select `name`, keyspace_id from name_user_vdx where `name` in ::__vals" => "select `name`, keyspace_id from name_user_vdx where `name` in ::name",
                                        "Table": "name_user_vdx",
                                        "Values": [
                                            "::name"
                                        ],
                                        "Variant": "IN",
                                        "Vindex": "user_index"
                                    },
                                    {
                                        "FieldQuery": "select 1, count(id) from `user` where 1 != 1",
                                        "Keyspace": {
                                            "Name": "user",
                                            "Sharded": true
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select 1, count(id) from `user` where `name` = 'a'",
                                        "Table": "`user`",
                                        "Variant": "ByDestination"
                                    }
                                ],
                                "Keyspace": {
                                    "Name": "user",
                                    "Sharded": true
                                },
                                "OperatorType": "VindexLookup",
                                "Values": [
                                    "'a'"
                                ],
                                "Variant": "Equal",
                                "Vindex": "name_user_map"
                            }
                        ],
                        "OperatorType": "Aggregate",
                        "Variant": "Scalar"
                    }
                ],
                "OperatorType": "Filter",
                "Predicate": "count(id) = 10",
                "ResultColumns": 1
            },
            "Original": "select 1 from user having count(id) = 10 and name = 'a'",
            "QueryType": "SELECT",
            "TablesUsed": [
                "user.user"
            ]
        }
```

**Comment by wangweicugw** [comment]
> > > @wangweicugw Nice stuff!
> > > Did you check that we have relevant end-to-end tests for this?
> > 
> > 
> > I haven't found relevant end-to-end tests for this; I will supplement them in go/test/endtoend/vtgate/queries/dml/dml_test.go.
> 
> you should also validate this in a unit test that shows what bind variables were sent down.

I have added some relevant test cases here.
go/vt/vtgate/engine/update_test.go
go/vt/vtgate/engine/delete_test.go

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:776]: The problem is in the `routing` engine primitive which assumes the bind variables always to have the prefix `__vals`. Only if we change that assumption can we fix this.
For now, we can keep this change

**Review by harshit-gangal — APPROVED** [review]

**Comment by wangweicugw** [comment]
@systay @harshit-gangal 
Will this PR be backported to release-18.0?

**Comment by harshit-gangal** [comment]
> @systay @harshit-gangal Will this PR be backported to release-18.0?

This is an enhancement PR therefore it does not qualify for backport

**Comment by harshit-gangal** [comment]
As this is an optimization PR, I think we should add benchmark showing if this reduces any memory allocations 
reference `go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go`

**Comment by wangweicugw** [comment]
> > @systay @harshit-gangal Will this PR be backported to release-18.0?
> 
> This is an enhancement PR therefore it does not qualify for backport

Alright, we are still using version 18.0.

**Comment by wangweicugw** [comment]
> As this is an optimization PR, I think we should add benchmark showing if this reduces any memory allocations reference `go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go`

I have attempted to add some benchmark tests, could you help me review them?

**Review by systay — APPROVED** [review]

**Pull request merged**